Source/Microsoft.PowerShell.Commands.Utility/SelectStringCommand.cs

```diff
@@ -0,0 +1,315 @@
+        [ValidateNotNullOrEmpty]
+        public string Encoding { get; set; }
+
+        [Parameter]
+        [ValidateNotNullOrEmpty]
+        public string[] Exclude { get; set; }
+
+        [ParameterAttribute]
+        [ValidateNotNullOrEmpty]
+        public string[] Include { get; set; }
+
+        [Parameter(
+            ValueFromPipeline = true,
+            Mandatory = true,
+            ParameterSetName = "Object")]
+        [AllowNull]
+        [AllowEmptyString]
+        public PSObject InputObject { get; set; }
+
+        [Parameter]
+        public SwitchParameter List { get; set; }
+
+        [ParameterAttribute(
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            ParameterSetName = "LiteralFile")]
+        [Alias("PSPath")]
+        public string[] LiteralPath {
+            get { return InternalPaths; }
+            set
+            {
+                AvoidWildcardExpansion = true;
+                InternalPaths = value;
+            }
+        }
+
+        bool AvoidWildcardExpansion;
+        string[] InternalPaths { get; set; }
+
+        [Parameter]
+        public SwitchParameter NotMatch { get; set; }
+
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            ParameterSetName = "File")]
+        public string[] Path {
+            get { return InternalPaths; }
+            set { InternalPaths = value; }
+        }
+
+        [Parameter(
+            Mandatory = true,
+            Position = 0)]
+        public string[] Pattern { get; set; }
+
+        [Parameter]
+        public SwitchParameter Quiet { get; set; }
+
+        [Parameter]
+        public SwitchParameter SimpleMatch { get; set; }
+
+        int _lineNumber = 1;
+        bool _matchedAtLeastOneItem;
+
+        protected override void ProcessRecord()
+        {
+            if (Quiet.IsPresent && _matchedAtLeastOneItem)
+            {
+                return;
+            }
+
+            if (Path != null)
+            {
+                MatchInFiles();
+            }
+            else if (InputObject != null)
+            {
+                MatchInputObject();
+            }
+        }
+
+        private void MatchInFiles()
+        {
+            foreach (string path in ResolvePaths())
+            {
+                MatchInLines(path, ReadLines(path));
+            }
+        }
+
+        private IEnumerable<string> ReadLines(string path)
+        {
+            return File.ReadAllLines(path, EncodingMapping.GetEncoding(Encoding));
+        }
+
+        private IEnumerable<string> ResolvePaths()
+        {
+            foreach (string path in InternalPaths)
+            {
+                CmdletProvider provider;
+                ProviderRuntime runtime = CreateProviderRuntime();
+                foreach (string resolvedPath in InvokeProvider.ChildItem.Globber.GetGlobbedProviderPaths(path, runtime, out provider))
```

Comment:
You could also use `SessionState.PathIntrinsics.GetResolvedPSPathFromPSPath(string[] path, ProviderRuntime runtime)`.

Just saying, might be more easy to read (path related) and safer (error/output handling is task of PathIntrinsics). The globbers work is more "basic".